### PR TITLE
Catch exception in group commit

### LINKED ISSUE #1440: Catch exception in group commit
### Describe the bug
In group commit `commit` function, the return value is true or false, denoting whether the group commit succeed or not. However, in the read data path, if there is an exception such as zkclient is closed, etc., an exception will still be thrown in group commit. This is inconsistent with group commit java doc, and also different from normal update zk operation, which catches all exception.

### To Reproduce
Zkclient closed will throw an exception in group commit.

### Expected behavior
Group commit catches all exceptions and return false.

### PR DESCRIPTION
**Pull request by zhangmeng916**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:

Fixed #1440 

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

In group commit commit function, the return value is true or false, denoting whether the group commit succeed or not. However, in the read data path, if there is an exception such as zkclient is closed, etc., an exception will still be thrown in group commit. This is inconsistent with group commit java doc, and also different from normal update zk operation, which catches all exception.
This PR catches all exception in commit function.

### Tests

- [X] The following is the result of the "mvn test" command on the appropriate module:

helix-core:
[WARNING] Tests run: 1210, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 3,947.76 s - in TestSuite
[INFO]
[INFO] Results:
[INFO]
[WARNING] Tests run: 1210, Failures: 0, Errors: 0, Skipped: 1
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:05 h
[INFO] Finished at: 2020-10-06T13:25:14-07:00
[INFO] ------------------------------------------------------------------------
~                                                                                       
helix-rest:
[INFO] Tests run: 169, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 141.451 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 169, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:25 min
[INFO] Finished at: 2020-10-06T13:43:14-07:00
[INFO] ----------------------------------------------

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/GroupCommit.java [hunk 1]
```diff
@@ -112,6 +112,9 @@ public boolean commit(BaseDataAccessor<ZNRecord> accessor, int options, String k
             merged = accessor.get(mergedKey, null, options);
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/GroupCommit.java:112]: I suggest covering all the accessor related calls inside the commit(). There should be 3 as I remember.

File: helix-core/src/main/java/org/apache/helix/GroupCommit.java [hunk 2]
```diff
@@ -112,6 +112,9 @@ public boolean commit(BaseDataAccessor<ZNRecord> accessor, int options, String k
...
           } catch (ZkNoNodeException e) {
             // OK.
+          } catch (Exception e) {
```
Comment (lei-xia) [inline helix-core/src/main/java/org/apache/helix/GroupCommit.java:115]: There are a few places that call DataAccessor.updateProperty but did not check the return results, we may want to double check whether it is appropriate (i.e, adding check there if it cares writing failure).

one example is:  HelixAdmin->addTypeToCustomizedStateConfig()

File: helix-core/src/main/java/org/apache/helix/GroupCommit.java (hunk 1 above)
Comment (lei-xia) [reply helix-core/src/main/java/org/apache/helix/GroupCommit.java:112]: What do you mean?

File: helix-core/src/main/java/org/apache/helix/GroupCommit.java (hunk 2 above)
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/GroupCommit.java:115]: Yep, I'll address them together with #1436

Comment (lei-xia) [reply helix-core/src/main/java/org/apache/helix/GroupCommit.java:115]: Can you search all existing use cases (no related to customized state) and make sure there is no assumption on old behavior?

**Review by jiajunwang — APPROVED** [review]
LGTM!

nit, it may also work just wrap the whole commit method in a huge try-catch block. But this one also works the same.

File: helix-core/src/main/java/org/apache/helix/GroupCommit.java (hunk 2 above)
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/GroupCommit.java:115]: This change only impacts group commit. As shown in the following, only current state and customized state use group commit. 
case CURRENTSTATES:
case CUSTOMIZEDSTATES:
        success = _groupCommit.commit(_baseDataAccessor, options, path, value.getRecord(), true);

However, as you mentioned, for normal update, there're cases that don't really check response. I'll take a look to see whether they're reasonable.

Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/GroupCommit.java:115]: Done. Actually no impact on other function.

**Comment by zhangmeng916** [comment]
This PR is ready to merge. Approved by @jiajunwang 
Final commit message:
Catch exception in group commit.

File: helix-core/src/main/java/org/apache/helix/GroupCommit.java [hunk 3]
```diff
@@ -139,14 +142,22 @@ public boolean commit(BaseDataAccessor<ZNRecord> accessor, int options, String k
           success = false;
           while (++retry <= MAX_RETRY && !success) {
             if (removeIfEmpty && merged.getMapFields().isEmpty()) {
-              success = accessor.remove(mergedKey, options);
+              try {
+                success = accessor.remove(mergedKey, options);
+              } catch (Exception e) {
+                success = false;
```
Comment (kaisun2000) [inline helix-core/src/main/java/org/apache/helix/GroupCommit.java:148]: Let us log this exception here.

File: helix-core/src/main/java/org/apache/helix/GroupCommit.java [hunk 4]
```diff
@@ -139,14 +142,22 @@ public boolean commit(BaseDataAccessor<ZNRecord> accessor, int options, String k
...
                 LOG.error("Fails to remove " + mergedKey + " from ZK, retry it!");
               } else {
                 LOG.info("Removed " + mergedKey);
               }
             } else {
-              success = accessor.set(mergedKey, merged, options);
+              try {
+                success = accessor.set(mergedKey, merged, options);
+              } catch (Exception e) {
+                success = false;
```
Comment (kaisun2000) [inline helix-core/src/main/java/org/apache/helix/GroupCommit.java:159]: log exception here.

File: helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java [hunk 5]
```diff
@@ -1315,8 +1315,11 @@ public void addTypeToCustomizedStateConfig(String clusterName, String type) {
     ZKHelixDataAccessor accessor =
         new ZKHelixDataAccessor(clusterName, new ZkBaseDataAccessor<ZNRecord>(_zkClient));
     PropertyKey.Builder keyBuilder = accessor.keyBuilder();
-    accessor.updateProperty(keyBuilder.customizedStateConfig(),
-        customizedStateConfigFromBuilder);
+    if(!accessor.updateProperty(keyBuilder.customizedStateConfig(),
```
Comment (kaisun2000) [inline helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java:1318]: This pull only deals with exception in group commit. How about the zkclient closed issue in callbackhandler?

File: helix-core/src/main/java/org/apache/helix/GroupCommit.java (hunk 3 above)
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/GroupCommit.java:148]: if you look at the next line. It's already logged there.

File: helix-core/src/main/java/org/apache/helix/GroupCommit.java (hunk 4 above)
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/GroupCommit.java:159]: It's logged next line.

File: helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java (hunk 5 above)
Comment (zhangmeng916) [reply helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java:1318]: The reason we need this PR is that we need to make sure updateProperty with group commit or non group commit have the same behavior. Currently non group commit update return error code, while group commit throws exception. This should not impact any callbackhandler existing logic.

File: helix-core/src/main/java/org/apache/helix/GroupCommit.java (hunk 3 above)
Comment (kaisun2000) [reply helix-core/src/main/java/org/apache/helix/GroupCommit.java:148]: The thing is that we did not log out the exception. At the time of investigation of production issue, you always want to know what is the exception thrown.

Current code actually "eat" the exception from IO (zk access). This is normally not a good idea.

how about something like this:

```
 try {
                success = accessor.remove(mergedKey, options);
  } catch (Exception e) {
                LOG.error (" failed to remove {} from zk, retry it", mergedkey, e);
                return false;
  }

 LOG.info( "removed {}", mergedkey);
```

**Review by jiajunwang — APPROVED** [review]
Reviewed after new code being updated.

**Pull request merged**